chapters/functions.tex

```diff
@@ -1166,22 +1166,107 @@ \subsection{Casting to Record}\label{casting-to-record}
+\end{nonnormative}
+
+The optional argument \lstinline!normallyConstant! of \lstinline!smoothOrder! defines that the function argument \lstinline!IDENT! is usually constant.
+
+\begin{nonnormative}
+A tool might check whether the actual argument to \lstinline!IDENT! is a parameter expression at the place where the function is called.
+If this is the case, the derivative of the function might be constructed under the assumption that the corresponding argument is constant, to enhance efficiency.
+Typically, a tool would generate at most two different derivative functions of a function:
+One, under the assumption that all \lstinline!normallyConstant! arguments are actually constant.
+And one, under the assumption that all input arguments are time varying.
+Based on the actual arguments of the function call either of the two derivative functions is used.
+
+This annotation is used by many functions of the \lstinline!Modelica.Fluid! library, such as\linebreak[4] \lstinline!Modelica.Fluid.Dissipation.PressureLoss.StraightPipe.dp_laminar_DP!, since geometric arguments to these functions are usually constant.
+\end{nonnormative}
+\end{semantics}
+\end{annotationdefinition}
+
+\begin{annotationdefinition}[derivative]
+\begin{synopsis}[grammar]\begin{lstlisting}
+"derivative" [ derivative-constraints ] "=" name
+
+derivative-constraints :
+   "(" derivative-constraint { "," derivative-constraint } ")"
+
+derivative-constraint :
+   "order" = UNSIGNED-NUMBER
+   | "noDerivative" = IDENT
+   | "zeroDerivative" = IDENT
+\end{lstlisting}\end{synopsis}
+\begin{semantics}
+This annotation has only an effect within a function declaration.
+
+The \lstinline!derivative! annotation can influence simulation time and accuracy, can be applied to both functions written in Modelica and to external functions, and may be used several times for the same function declaration.
+
+Each use of the \lstinline!derivative! annotation points to another \firstuse{derivative-function}\index{derivative-function} that expresses a derivative of the declared function, and the annotation can state that it is only valid under certain restrictions on the input arguments.
+These restrictions are defined using the optional attributes \lstinline!order!, \lstinline!noDerivative!, and \lstinline!zeroDerivative!.
+The \lstinline!order! may be specified at most once for each \lstinline!derivative! annotation, must be at least 1, and defaults to 1.
+Specifying \lstinline!order! is only considered a restriction if $\text{\lstinline!order!} > 1$.
+
+For details abouts using the \lstinline!derivative! annotation, see \cref{using-the-derivative-annotation}.
+\end{semantics}
+\end{annotationdefinition}
+
+\begin{annotationdefinition}[inverse]
+\begin{synopsis}[grammar]\begin{lstlisting}
+"inverse" "(" function-inverse { "," function-inverse } ")"
+
+function-inverse :
+   IDENT "=" name "(" IDENT { "," IDENT } ")"
```

Comment:
This grammar is a new change and I think it is over-specifying it, and breaking backward compatibility.
```suggestion
   IDENT "=" type-specifier function-call-args"
```
There are a number of reasons for this:

- Type-specifier allows leading "." for global names, which is clearly good.
- It allows named arguments in the call as well (good for readability).
- It allows literal values as well.